File: contest/LeetCode_Weekly_Contest_201/Maximum_Number_of_Non-Overlapping_Subarrays_With_Sum_Equals_Target.java
class Solution {
    public int maxNonOverlapping(int[] nums, int target) {
        int n = nums.length;
        HashMap<Integer, Integer> pos = new HashMap<Integer, Integer>();

        int[] dp = new int[n+1];
        for (int i = 0; i <= n; ++i) dp[i] = 0;
        pos.put(0, n);

        int suffix = 0, ret = 0;

        for (int i = n-1; i >= 0; --i) {
            dp[i] = Math.max(dp[i], dp[i+1]);
            suffix += nums[i];
            int v = suffix - target;
            if (pos.containsKey(v)) { 
                int p = pos.get(v);
                dp[i] = Math.max(dp[i], dp[p] + 1);
            }
            pos.put(suffix, i);
            ret = Math.max(ret, dp[i]);
        }
        return ret;
    }
}

File: contest/LeetCode_Weekly_Contest_201/Maximum_Number_of_Non-Overlapping_Subarrays_With_Sum_Equals_Target.py
class Solution(object):
    def maxNonOverlapping(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        
        n = len(nums)
        dp = [0 for i in xrange(n+1)]
        pos = {0: n}
        suffix = 0
        ret = 0
        for i in xrange(n-1, -1, -1):
            suffix += nums[i]
            v = suffix - target
            dp[i] = max(dp[i], dp[i+1])
            if v in pos:
                p = pos[v]
                dp[i] = max(dp[i], dp[p]+1)
            pos[suffix] = i
            ret = max(ret, dp[i])
        return ret
        
